R/duplicated.R

```diff
@@ -140,14 +140,25 @@ uniqueN = function(x, by = if (is.list(x)) seq_along(x) else NULL, na.rm=FALSE)
     by = key(x)
     stop(error_oldUniqueByKey)
   }
+  #verbose = getOption("datatable.verbose")
   if (is.null(x)) return(0L)
   if (!is.atomic(x) && !is.data.frame(x))
     stop("x must be an atomic vector or data.frames/data.tables")
   if (is.atomic(x)) {
-    if (is.logical(x)) return(.Call(CuniqueNlogical, x, na.rm=na.rm))
+    if (is.logical(x)) {
+    #  if (verbose) cat("uniqueN detected logical type, using optimised C uniqueNlogical\n")
+      return(.Call(CuniqueNlogical, x, na.rm=na.rm))
+    }
+    escape_n = 1e3L #getOption("datatable.escapeUniqueN",1e7L)
+    if (length(x) <= escape_n) {
+      #if (verbose) cat("uniqueN detected atomic type of length ", length(x), " which is small enough to use base R ", if (na.rm) "na.omit+" else "", "duplicated\n",sep="")
+      if (na.rm) x = na.omit(x)
+      return(length(x)-sum(duplicated(x)))
```

Comment:
Probably slightly quicker to not `na.omit` but just do:

```r
return(length(x) - sum(duplicated(x)) - {na.rm && anyNA(x)})
```

Reply:
I am not getting how that would work

Reply:
If `na.rm = FALSE` then my version is just `length(x) - sum(duplicated(x)) - 0`. 

If `na.rm = TRUE` then it will be `length(x) - sum(duplicated(x)) - 1 * anyNA(x)`.

Reply:
(key insight being that `uniqueN` can differ by at most 1 for `na.rm` vs `!na.rm`)

Reply:
nice one!
it seems that it might eventually differ by 2
```
> x=c(1,2,NA,2,3,NaN,NaN,NA)
> uniqueN(x)
[1] 5
> uniqueN(x, na.rm=T)
[1] 3
> uniqueN(na.omit(x))
[1] 3
```